devices/src/virtio/vsock/csm/txbuf.rs

```diff
@@ -0,0 +1,289 @@
+// Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use std::io::{ErrorKind, Write};
+use std::mem;
+use std::num::Wrapping;
+
+use super::defs;
+use super::{Error, Result};
+
+/// A simple ring-buffer implementation, used by vsock connections to buffer TX (guest -> host) data.
+/// Memory for this buffer is allocated lazily, since buffering will only be needed when the host
+/// can't read fast enough.
+///
+pub struct TxBuf {
```

Comment:
I agree this could be turned in to a general purpose ring buffer, but there are few things to iron-out first:

- the `push()` behavior: push all or nothing; this check is currently made here, instead of at the buffer user's end, which might not be suitable for all use cases; it's fine here, since data needs to be stored fully, or the whole connection gets dropped, as insufficient buffer space implies a misbehaving peer;
- the lazy uninit memory allocation might, again, not be generally desirable;
- the non-configurable buffer size.

While I agree that all of the above can be implemented as user-configurable, my reasoning was that designing that flexible interface could be postponed until the moment we need another ring buffer somewhere. So, lazy allocation of coding resources, I guess? I'm not particularly fond of this approach, but it seemed like an acceptable compromise at the time.

Reply:
I am ok with leaving it as it is.